# FMI-Kumpula | Solar Power

In [1]:
#Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import plotly.graph_objs as go
import plotly.express as ex
from plotly.offline import iplot, init_notebook_mode
import cufflinks
cufflinks.go_offline(connected = True)
init_notebook_mode(connected = True)

# 1. Read and Preprocessing Data

In [2]:
#Import data and print samples
X = pd.read_csv('FMI-Kumpula.csv')
X.sample(10)

request_id           forecast_time  power_output_w  power_output_f0_w  \
200        4405  2020-10-11 12:00:00+03           901.8               75.6   
261        4405  2020-10-14 01:00:00+03             NaN                NaN   
240        4405  2020-10-13 04:00:00+03             NaN                NaN   
107        4447  2020-10-15 06:00:00+03             NaN                NaN   
264        4357  2020-10-06 01:00:00+03             NaN                NaN   
297        4357  2020-10-07 10:00:00+03          1676.5              287.8   
65         4492  2020-10-22 09:00:00+03             0.0                NaN   
21         4492  2020-10-20 13:00:00+03          7869.9             6959.5   
244        4405  2020-10-13 08:00:00+03             NaN                NaN   
228        4405  2020-10-12 16:00:00+03          2305.8              867.8   

     power_output_f10_w  power_output_f25_w  power_output_f50_w  \
200               117.2               189.8               313.2   
261                 NaN                 NaN                 NaN   
240                 NaN                 NaN                 NaN   
107                 NaN                 NaN                 NaN   
264                 NaN                 NaN                 NaN   
297               449.9               689.1              1114.7   
65                  NaN                 NaN                 NaN   
21               7757.5              7822.9              7870.0   
244                 NaN                 NaN                 NaN   
228              1269.4              1703.3              2201.7   

     power_output_f75_w  power_output_f90_w  power_output_f100_w  ...  \
200               541.5               822.3               1664.6  ...   
261                 NaN                 NaN                  NaN  ...   
240                 NaN                 NaN                  NaN  ...   
107                 NaN                 NaN                  NaN  ...   
264                 NaN                 NaN                  NaN  ...   
297              2057.1              2659.1               3127.3  ...   
65                  NaN                 NaN                  NaN  ...   
21               7914.5              7932.8               7974.6  ...   
244                 NaN                 NaN                  NaN  ...   
228              2904.5              3476.5               3903.2  ...   

     cloud_cover_total  cloud_cover_high  cloud_cover_medium  cloud_cover_low  \
200              100.0             100.0                88.9             99.9   
261               18.9               0.0                 0.0             18.9   
240              100.0             100.0               100.0             53.4   
107               27.5               0.0                 0.0             27.5   
264              100.0              24.1                 0.1            100.0   
297               99.4               0.0                99.4              8.1   
65               100.0             100.0               100.0            100.0   
21                 0.4               0.0                 0.0              0.3   
244              100.0               0.0               100.0             99.7   
228              100.0              99.8               100.0             54.4   

     system_radiation_global_wm2  system_radiation_direct_wm2  \
200                        59.14                         3.91   
261                          NaN                          NaN   
240                          NaN                          NaN   
107                          NaN                          NaN   
264                          NaN                          NaN   
297                        98.16                        11.29   
65                          0.13                         0.00   
21                        380.34                       113.65   
244                          NaN                          NaN   
228                       128.49                         6.92 

In [3]:
print('------------- COLUMN TYPE -------------')
print(X.dtypes)

------------- COLUMN TYPE -------------
request_id                        int64
forecast_time                    object
power_output_w                  float64
power_output_f0_w               float64
power_output_f10_w              float64
power_output_f25_w              float64
power_output_f50_w              float64
power_output_f75_w              float64
power_output_f90_w              float64
power_output_f100_w             float64
system_temperature_c            float64
nominal_output_efficiency       float64
air_temperature_c               float64
cloud_cover_total               float64
cloud_cover_high                float64
cloud_cover_medium              float64
cloud_cover_low                 float64
system_radiation_global_wm2     float64
system_radiation_direct_wm2     float64
system_radiation_diffuse_wm2    float64
radiation_global_wm2            float64
radiation_direct_wm2            float64
radiation_diffuse_wm2           float64
dtype: object


In [4]:
#Change column forecast_time to date time format
X['forecast_time'] = pd.to_datetime(X['forecast_time'], format = "%Y-%m-%d %H:%M:%S")
print('--------------------- COLUMN TYPE AFTER CHANGING --------------------')
print(X.dtypes)

--------------------- COLUMN TYPE AFTER CHANGING --------------------
request_id                                                      int64
forecast_time                   datetime64[ns, pytz.FixedOffset(180)]
power_output_w                                                float64
power_output_f0_w                                             float64
power_output_f10_w                                            float64
power_output_f25_w                                            float64
power_output_f50_w                                            float64
power_output_f75_w                                            float64
power_output_f90_w                                            float64
power_output_f100_w                                           float64
system_temperature_c                                          float64
nominal_output_efficiency                                     float64
air_temperature_c                                             float64
cloud_cover_total   

In [5]:
#The request_id column seems to hold no information
#Remove reques_id column and set forecast_time to be the index column
X.drop('request_id', inplace = True, axis = 1)
X.set_index(['forecast_time'], inplace = True)
X.sample(5)

power_output_w  power_output_f0_w  \
forecast_time                                                  
2020-10-21 23:00:00+03:00             NaN                NaN   
2020-10-15 06:00:00+03:00             NaN                NaN   
2020-10-22 00:00:00+03:00             NaN                NaN   
2020-10-15 23:00:00+03:00             NaN                NaN   
2020-10-07 19:00:00+03:00             0.0                0.0   

                           power_output_f10_w  power_output_f25_w  \
forecast_time                                                       
2020-10-21 23:00:00+03:00                 NaN                 NaN   
2020-10-15 06:00:00+03:00                 NaN                 NaN   
2020-10-22 00:00:00+03:00                 NaN                 NaN   
2020-10-15 23:00:00+03:00                 NaN                 NaN   
2020-10-07 19:00:00+03:00                 0.0                 0.0   

                           power_output_f50_w  power_output_f75_w  \
forecast_time                                                       
2020-10-21 23:00:00+03:00                 NaN                 NaN   
2020-10-15 06:00:00+03:00                 NaN                 NaN   
2020-10-22 00:00:00+03:00                 NaN                 NaN   
2020-10-15 23:00:00+03:00                 NaN                 NaN   
2020-10-07 19:00:00+03:00                 0.0                 0.0   

                           power_output_f90_w  power_output_f100_w  \
forecast_time                                                        
2020-10-21 23:00:00+03:00                 NaN                  NaN   
2020-10-15 06:00:00+03:00                 NaN                  NaN   
2020-10-22 00:00:00+03:00                 NaN                  NaN   
2020-10-15 23:00:00+03:00                 NaN                  NaN   
2020-10-07 19:00:00+03:00                 0.0                  0.0   

                           system_temperature_c  nominal_output_efficiency  \
forecast_time                                                                
2020-10-21 23:00:00+03:00                  7.10                        NaN   
2020-10-15 06:00:00+03:00                  5.10                        NaN   
2020-10-22 00:00:00+03:00                  7.50                        NaN   
2020-10-15 23:00:00+03:00                  3.45                        NaN   
2020-10-07 19:00:00+03:00                 13.96                        0.0   

                           ...  cloud_cover_total  cloud_cover_high  \
forecast_time              ...                                        
2020-10-21 23:00:00+03:00  ...              100.0             100.0   
2020-10-15 06:00:00+03:00  ...               27.5               0.0   
2020-10-22 00:00:00+03:00  ...              100.0             100.0   
2020-10-15 23:00:00+03:00  ...              100.0             100.0   
2020-10-07 19:00:00+03:00  ...              100.0              47.3   

                           cloud_cover_medium  cloud_cover_low  \
forecast_time                                                    
2020-10-21 23:00:00+03:00                98.8            100.0   
2020-10-15 06:00:00+03:00                 0.0             27.5   
2020-10-22 00:00:00+03:00                68.2            100.0   
2020-10-15 23:00:00+03:00                79.9              0.1   
2020-10-07 19:00:00+03:00               100.0              0.1   

                           system_radiation_global_wm2  \
forecast_time                                            
2020-10-21 23:00:00+03:00                          NaN   
2020-10-15 06:00:00+03:00                          NaN   
2020-10-22 00:00:00+03:00                          NaN   
2020-10-15 23:00:00+03:00                          NaN   
2020-10-07 19:00:00+03:00                          0.5   

                           system_radiation_direct_wm2  \
forecast_time                                            
2020-10-21 23:00:00+03:00                          NaN   
2020-10-15 06:00:00+03:00     

## 2. Explore data

In [6]:
print('Shape =', X.shape)
print()
print('NaN value at each column:')
print(X.isna().sum())

Shape = (330, 21)

NaN value at each column:
power_output_w                  195
power_output_f0_w               213
power_output_f10_w              213
power_output_f25_w              213
power_output_f50_w              213
power_output_f75_w              213
power_output_f90_w              213
power_output_f100_w             213
system_temperature_c              0
nominal_output_efficiency       195
air_temperature_c                 0
cloud_cover_total                 0
cloud_cover_high                  0
cloud_cover_medium                0
cloud_cover_low                   0
system_radiation_global_wm2     195
system_radiation_direct_wm2     195
system_radiation_diffuse_wm2    195
radiation_global_wm2              0
radiation_direct_wm2              0
radiation_diffuse_wm2             0
dtype: int64


There are total 330 rows, but quite a number of rows have missing values.

The time is not in the right order. It needs to be sorted.

In [7]:
X = X.sort_values('forecast_time')

In [8]:
X.describe().round(3)

power_output_w  power_output_f0_w  power_output_f10_w  \
count         135.000            117.000             117.000   
mean         2620.769           1005.557            1425.832   
std          2460.875           1504.317            1699.427   
min             0.000              0.000               0.000   
25%           427.500            107.800             171.400   
50%          2020.600            407.400             886.100   
75%          4533.800           1222.800            1936.500   
max          8327.500           6959.500            7757.500   

       power_output_f25_w  power_output_f50_w  power_output_f75_w  \
count             117.000             117.000             117.000   
mean             1928.747            2533.238            3122.935   
std              1937.933            2231.488            2542.896   
min                 0.000               0.000               0.000   
25%               258.000             399.000             541.500   
50%              1508.900            2201.700            2793.900   
75%              2918.100            4058.400            5349.500   
max              7822.900            8157.700            8295.600   

       power_output_f90_w  power_output_f100_w  system_temperature_c  \
count             117.000              117.000               330.000   
mean             3479.615             3718.249                 9.749   
std              2715.003             2811.739                 5.165   
min                 0.000                0.000                -0.600   
25%               716.800              903.600                 6.200   
50%              3395.200             3779.500                 9.175   
75%              5920.900             6484.400                13.435   
max              8354.300             8384.900                21.830   

       nominal_output_efficiency  ...  cloud_cover_total  cloud_cover_high  \
count                    135.000  ...            330.000           330.000   
mean                       0.738  ...             72.670            41.888   
std                        0.315  ...             36.554            44.123   
min                        0.000  ...              0.000             0.000   
25%                        0.629  ...             43.200             0.000   
50%                        0.888  ...             96.400            17.250   
75%                        0.970  ...            100.000            96.525   
max                        1.034  ...            100.000           100.000   

       cloud_cover_medium  cloud_cover_low  system_radiation_global_wm2  \
count             330.000          330.000                      135.000   
mean               31.789           51.182                      139.237   
std                43.148           41.459                      119.367   
min                 0.000            0.000                        0.130   
25%                 0.000            5.650                       34.005   
50%                 0.300           47.000                      114.220   
75%                90.700           98.600                      235.005   
max               100.000          100.000                      407.280   

       system_radiation_direct_wm2  system_radiation_diffuse_wm2  \
count                      135.000                       135.000   
mean                        24.556                       114.352   
std                         34.203                        89.910   
min                          0.000                         0.130   
25%                          0.035                        33.410   
50%                          6.920                       102.500   
75%                         31.895                       199.915   
max                        132.800                       362.730   

       radiation_global_wm2  radiation_direct_wm2  radiation_diffuse_wm2  
count               330.000               330.000                330.000  
mean             

## 3. Visualizing Data